Update v12 sharding.md example code (#98)
AnIdiotsGuide/discordjs-bot-guide: understanding/sharding.md

There are additional difficulties when sharding a bot that add complexity to your code \(one of the reasons you shouldn't shard too early\). You do not need to worry about sharding until your bot hits around 2,400 guilds. YOU MUST SHARD by the time you hit 2,500 guilds, however.

## Sharding Styles

There are two styles of sharding that we'll be discussing: [internal sharding](#internal-sharding) and [traditional sharding](#traditional-sharding). Each of these sharding styles holds benefits depending on your situation.

## Internal Sharding

`internal` sharding is the method by which a bot's code creates multiple shard connections to the Discord API *within a single process*. This means that all the guilds, channels, and users on one shard will be available to another shard via a direct call (e.g. `client.guilds.cache.get('GUILD_ID')`). Due to the large memory size the single bot process will grow to using this style of sharding, it is not ideal for bots with many guilds.

### Internal Sharding Caveats

* Internally sharded bots often gain very marginal performance increase and will use *more memory*.

> * If you're using any sort of database or connection, multiple shards may cause issues with multiple processes connecting to a single end point.

Reviewer: This is true for the sharding manager, but not for internal sharding, so we don't need that warning here

// also set this to a number.
const client = new Discord.client({ shardCount: 'auto'});
```

## Traditional Sharding

`traditional` sharding is the method by which a bot's code spawns individual child processes via a main shard manager process, each child process being one shard of the bot. When using this style of sharding, guilds, channels, and users on one shard will *not* be available to another via direct call (e.g. `client.guilds.cache.get('GUILD_ID')`) because each shard is in a separate process.

This style of sharding is ideal for larger bots, or bots that need to be scalable to allow for future growth. The rest of this page will discuss [how to make use of traditional sharding](#example-sharding-manager-code) and [how to share information between shards](#sharding-information-between-shards).

To learn how to make use of this, read on!

### Traditional Sharding Caveats

* Collections do not cache data from all shards, so you can't grab data from a guild in another shard easily.
* In order to do anything across shards you need to worry about using [`fetchClientValues`](#fetchclientvalues) and [`broadcastEval`](#broadcasteval) \(Examples and explanation below\).
And again:

> * Traditionally sharded bots often gain very marginal performance increase and might even use *more memory* due to using more node processes.

Reviewer: So both internal sharding and traditional sharding use more memory, so the question is, what are the *advantages* of using internal, or traditional, sharding? Caveats are one thing, but why would one do it other than to just satisfy Discord's requirements?

Reviewer 2: Internal sharding comes with near to no memory overhead, whereas separate processes still have to load all the dependencies to memory, which in large bots can be 20-50mb.

Therefore, this is a valid caveat of traditional sharding, specially when they stack up.

Example:
```javascript
/*
    Example result of fetchClientValues() on a bot with 4,300 guilds split across 4 shards.
    Assume this is being executed on shard 0, the first shard.
*/

// If we just get our client.guilds.cache.size, it will return
// only the number of guilds on the shard this is being run on.
console.log(client.guilds.cache.size);
// 1050

// If we would like to get our client.guilds.cache.size from all
// of our shards, we must make use of fetchClientValues().
const res = await client.shard.fetchClientValues('guilds.cache.size');
console.log(res);
// 	[
//		1050,	// shard 0
//		1100,	// shard 1
//		1075,	// shard 2
//		1075	// shard 3
//	]

````

Let's say you want to do something like get your total server count - In a non-sharded environment, this would be as simple as getting the `client.guilds.cache.size`. However, in this case `client.guilds.cache.size` will not return the total servers your bot is in. Instead it returns only the total number of servers *on this shard*, like in the first part of the example above.

> 

Reviewer: I feel it's a bit redundant to have both the comments, and this paragraph, showing the same thing, and then below to show a real-life example of using the code above. This can be made more concise by removing the comments and just having the text.

Example:
```javascript
/*
    Example of result of broadcastEval() on a bot with 4 servers split across 2 shards.
    Assume this is being executed on shard 0, the first shard.
*/

// If we just map our guilds' members.cache.size, it will return
// only the mapped members.size of the shard this is being run on.
console.log(client.guilds.cache.map((guild) => guild.members.cache.size));
//		[
//			30,
//			25
//		],

// If we would like to map our guilds' members.cache.size from our
// servers on all of our shards, we must make use of broadcastEval().
// Remember, this runs in the context of the client, so we refer to the
// Client using "this".
const res = await client.shard.broadcastEval('this.guilds.cache.map((guild) => guild.members.cache.size)');
console.log(res);
// 	[
//		[	// shard 0
//			30,
//			25
//		],
//		[	// shard 1
//			55,
//			10
//		]
// 	]

````

> Say you want to get a guild from your client. In a non-sharded environment, you would simply use `client.guilds.cache.get('ID')` or something of that nature and then carry on with your code. In this case however, it is possible that the guild you're trying to get *is not present on the shard*. In order to get the guild for use, you would then need to fetch it from whatever shard it is present on using `broadcastEval()`.

Reviewer: This is, again, redundant because it says the same thing as the code samples above.

Here's an example of a function that uses `broadcastEval()` to get a single guild no matter what shard it is present on:

```javascript
/*
  	Example by ZiNc#2032
  	The following code fetches a single guild from across shards.
    
    NOTE: Fetched guild's properties such as "Guild.members.cache" and "Guild.roles.cache" will
    not be Managers or Collections; these properties will be arrays of snowflake IDs.
  	
  	discord.js version 12.x
  	client = new discordjs.Client()

  	returns Guild
*/

const getServer = async (guildID) => {
    // try to get guild from all the shards
    const req = await client.shard.broadcastEval(`this.guilds.cache.get("${guildID}")`);

    // return Guild or null if not found
    return req.find(res => !!res) || null;
}
```

{% hint style="info" %}

> `broadcastEval()` will only return basic javascript objects, and will not return Discord.js structures such as Collections, Guild objects, Member object, and the likes. You will need to interact directly with the API or build these structures yourself after getting a response back from your `broadcastEval()`.

Reviewer: I've love to see an example structure, as we've seen above with all the other code.

/*
    The following code goes into it's own file, and you run this file
    instead of your main bot file.
*/

// Include discord.js ShardingManger
const { ShardingManager } = require('discord.js');

// Create your ShardingManger instance
const manager = new ShardingManager('./YOUR_BOT_FILE_NAME.js', {
    // for ShardingManager options see:
    // https://discord.js.org/#/docs/main/v12/class/ShardingManager
    totalShards: 'auto',
    token: 'YOUR_TOKEN_GOES_HERE'
});

// Spawn your shards
manager.spawn();

// Emitted when a shard is created
manager.on('shardCreate', (shard) => console.log(`Shard ${shard.id} launched`));
```

## Sharing Information Between Shards

Remember how we were talking about sharding being a method of "splitting" the bot into multiple instances of itself? Because of this, things like your adding your total guilds or getting a specific guild are not as simple as they were before. We must now use either [`fetchClientValues`](#fetchclientvalues) or [`broadcastEval`](#broadcasteval) to get information from across shards.

These two functions are your go-to for getting any information from other shards, so get familiar with them!

### FetchClientValues

> [`fetchClientValues`](https://discord.js.org/#/docs/main/v12/class/ShardClientUtil?scrollTo=fetchClientValues) gets Client properties from all shards. This is what you should use when you would like to get any of the nested properties of the Client, such as `guilds.size` or `uptime`. It's useful for getting things like Collection sizes, basic client properties, and unprocessed information about the client.

Reviewer 3: `guilds.size` ->  `guilds.cache.size`

  	returns number
*/

const getServerCount = async () => {
    // get guild collection size from all the shards
    const req = await client.shard.fetchClientValues('guilds.cache.size');

    // return the added value
    return req.reduce((p, n) => p + n, 0);
}
```

{% hint style="info" %}
`fetchClientValues()` does not allow you to make use of javascript methods or client methods to get or process information before returning it. It only allows you to get information from client properties.
{% endhint %}

### BroacastEval

[`broadcastEval`](https://discord.js.org/#/docs/main/v12/class/ShardClientUtil?scrollTo=broadcastEval) evaluates the input in the context of each shard's Client(s). This is what you should use when you want to execute a method or process data on a shard and return the result. It's useful for getting information that isn't available through client properties and must instead be retrieved through the use of methods.

Example:
```javascript
/*
    Example of result of broadcastEval() on a bot with 4 servers split across 2 shards.
    Assume this is being executed on shard 0, the first shard.
*/

// If we just map our guilds' members.cache.size, it will return

> // only the mapped members.size of the shard this is being run on.

Reviewer 3: `members.size` -> `members.cache.size`

You do not need to worry about sharding until your bot hits around 2,400 guilds. YOU MUST SHARD before you hit 2,500 guilds, however.
{% endhint %}

## Sharding Styles

There are two styles of sharding that we'll be discussing: [`internal` sharding](#internal-sharding) and [`traditional` sharding](#traditional-sharding). Each of these sharding styles holds benefits depending on your situation.

### Internal Sharding

> `internal` sharding is the method by which a bot's code creates multiple shard connections to the Discord API *within a single process*. This means that all the guilds, channels, and users on one shard will be available to another shard via a direct call (e.g. `client.guilds.cache.get('GUILD_ID')`) because all of the shards share the same process.

Reviewer 2: "within a single process" and "because all of the shards share the same process" in the same paragraph feels like redundancy.

Perhaps change the latter into something else, or just remove it?

> Because an internally sharded bot is run in one process, it is not ideal for bots with many guilds due to the large memory size the process will grow to. If you would like to use internal sharding, adjust the Client options in your main bot file where you define your client.

Reviewer 2: I think the first sentence could be changed to "Because of the previously mentioned design", as the previous paragraph gives it the context it needs, but I'm not sure, thoughts @eslachance?

Also "If you would like to use internal sharding" can be changed to "If you would like to use this".

Reviewer 2: We also have a "Traditional Sharding Caveats", and this paragraph would fit for an "Internal Sharding Caveats", so this could be moved down right before "Traditional Sharding" and jump into the example. This will invalidate the first part of the previous review.

Here is an example of how to make use of internal sharding:

```javascript
/*
    The following code goes into your main bot file.
*/

// Include discord.js ShardingManger
const Discord = require("discord.js");

// When we define our client, we include the property "shardCount"
// and set it to 'auto' to allow the client to automatically create
// the correct number of shards.
// If you would like to have a different number of shards, you may
// also set this to a number.
const client = new Discord.client({ shardCount: 'auto'});
```

### Traditional Sharding

`traditional` sharding is the method by which a bot's code spawns individual child processes via a main shard manager process, each child process being one shard of the bot. When using this style of sharding, guilds, channels, and users on one shard will *not* be available to another via direct call (e.g. `client.guilds.cache.get('GUILD_ID')`) because each shard is in a separate process.

> This style of sharding is ideal for larger bots, or bots that need to be scalable to allow for future growth. The rest of this page will discuss [how to make use of traditional sharding](#example-sharding-manager-code) and [how to share information between shards](#sharding-information-between-shards) as information is not readily available between shards.

Reviewer 2: "as information is not readily available between shards." is duplicated with linked section, which has "Information is not readily available between shards.".

To learn how to make use of traditional sharding, read on!

> ## Traditional Sharding Caveats

Reviewer 2: This and the following H2 (`##`) titles are part of the H3 (`###`) "Traditional Sharding" title. Maybe change them to H4 `####`?

These two functions are your go-to for getting any information from other shards, so get familiar with them!

## FetchClientValues

[`fetchClientValues`](https://discord.js.org/#/docs/main/v12/class/ShardClientUtil?scrollTo=fetchClientValues) gets Client properties from all shards. This is what you should use when you would like to get any of the nested properties of the Client, such as `guilds.size` or `uptime`. It's useful for getting things like Collection sizes, basic client properties, and unprocessed information about the client.

Example:
```javascript
/*
    Example result of fetchClientValues() on a bot with 4,300 guilds split across 4 shards.
    Assume this is being executed on shard 0, the first shard.
*/

// If we just get our client.guilds.cache.size, it will return
// only the number of guilds on the shard this is being run on.

> console.log('client.guilds.cache.size');
> // 1050

Reviewer 2: Did you mean to do this?
```suggestion
console.log(client.guilds.cache.size);
// 1050
```

// If we would like to get our client.guilds.cache.size from all
// of our shards, we must make use of fetchClientValues().
const res = await client.shard.fetchClientValues('guilds.cache.size');
console.log(res);
// 	[
//		1050,	// shard 0
//		1100,	// shard 1
//		1075,	// shard 2
//		1075	// shard 3
//	]

````

> Let's say you want to do something like get your total server count - In a non-sharded environment, this would be as simple as getting the `client.guilds.cache.size`. However in a sharded environment, `client.guilds.cache.size` will return not the total servers your bot is in. Instead it returns only the total number of servers *on this shard*, like in the first part of the example above.

Reviewer 2: "Non-sharded environment" and "sharded environment" kind of feel like `if (false) {} else if (true) {}`.

Maybe change the latter to "However, in this case `client.guilds.cache.size` will...".

Example:
```javascript
/*
    Example of result of broadcastEval() on a bot with 4 servers split across 2 shards.
    Assume this is being executed on shard 0, the first shard.
*/

// If we just map our guilds' members.cache.size, it will return
// only the mapped members.size of the shard this is being run on.
console.log(client.guilds.cache.map((guild) => guild.members.cache.size));
//		[
//			30,
//			25
//		],

// If we would like to map our guilds' members.cache.size from our
// servers on all of our shards, we must make use of broadcastEval().
// Remember, this runs in the context of the client, so we refer to the
// Client using "this".
const res = await client.shard.broadcastEval('this.guilds.cache.map((guild) => guild.members.cache.size)');
console.log(res);
// 	[
//		[	// shard 0
//			30,
//			25
//		],
//		[	// shard 1
//			55,
//			10
//		]
// 	]

````

> Say you want to get a guild from your client. In a non-sharded environment, you would simply use `client.guilds.cache.get('ID')` or something of that nature and then carry on with your code. In a sharded environment however, it is possible that the guild you're trying to get *is not present on the shard*. In order to get the guild for use, you would then need to fetch it from whatever shard it is present on using `broadcastEval()`.

Reviewer 2: Same thing as before (redundancy of "non-sharded environment" and "sharded environment" in the same paragraph).

Here's an example of a function that uses `broadcastEval()` to get a single guild no matter what shard it is present on:

```javascript
/*
  	Example by ZiNc#2032
  	The following code fetches a single guild from across shards.
    
    NOTE: Fetched guild's properties such as "Guild.members.cache" and "Guild.roles.cache" will
    not be Managers or Collections; these properties will be arrays of snowflake IDs.
  	
  	discord.js version 12.x
  	client = new discordjs.Client()

  	returns Guild
*/

const getServer = async (guildID) => {
    // try to get guild from all the shards
    const req = await client.shard.broadcastEval(`this.guilds.cache.get("${guildID}")`);

>     // return non-null response or false if not found
>     return (req.find((res) => !!res) || false);

Reviewer 2: Not going to lie, but returning `false` in a method that may return an object may confuse more experienced developers when people who use this code ask for support.

`Array.prototype.find` returns `undefined`, not `null`.

```suggestion
    // return an object or null if not found
    return req.find(Boolean) || null;
```

Alternatively (also works, though, but uses more syntax):

```suggestion
    // return an object or null if not found
    return req.find(res => !!res) || null;
```

There are additional difficulties when sharding a bot that add complexity to your code \(one of the reasons you shouldn't shard too early\).

* Collections do not cache data from all shards, so you can't grab data from a guild in another shard easily.
* In order to do anything across shards you need to worry about using `broadcastEval` and such \(Examples and explanation below\).
* Sharded bots often gain very marginal performance increase and might even use _more_ memory due to using more node processes.
* If you're using any sort of database or connection, multiple shards may cause issues with multiple processes connecting to a single end point.

## Example Sharding Manager Code

```javascript
/*
    The following code goes into it's own file, and you run this file
    instead of your main bot file.
*/

// Include discord.js ShardingManger
const { ShardingManager } = require('discord.js');

// Create your ShardingManger instance
const manager = new ShardingManager('./YOUR_BOT_FILE_NAME.js', {
    // for ShardingManager options see:

>     // https://discord.js.org/#/docs/main/master/class/ShardingManager

Reviewer 2: This should target `stable`, as not everyone targets master.
```suggestion
    // https://discord.js.org/#/docs/main/stable/class/ShardingManager
```

Reviewer 2: Same thing for all other instances.

    totalShards: 'auto',
    token: 'YOUR_TOKEN_GOES_HERE'
});

// Spawn your shards
manager.spawn();

// Emitted when a shard is created
manager.on('shardCreate', (shard) => console.log(`Shard ${shard.id} launched`));
```

## Sharing Information Between Shards

{% hint style="info" %}
Information is not readily available between shards. In order to get or share information across shards, you will need to make use of either `fetchClientValues()` or `broadcastEval()`.
{% endhint %}

> Remember how we were talking about sharding being a method of "splitting" the bot into multiple instances of itself? Because your sharded bot is now in separate, individual instances, things like your adding your total guilds or getting a specific guild are not as simple as they were before. We must now use either [`fetchClientValues`](https://discord.js.org/#/docs/main/master/class/ShardClientUtil?scrollTo=fetchClientValues) or [`broadcastEval`](https://discord.js.org/#/docs/main/master/class/ShardClientUtil?scrollTo=broadcastEval) to get information from across shards.

Reviewer 2: Same here:

```suggestion
Remember how we were talking about sharding being a method of "splitting" the bot into multiple instances of itself? Because your sharded bot is now in separate, individual instances, things like your adding your total guilds or getting a specific guild are not as simple as they were before. We must now use either [`fetchClientValues`](https://discord.js.org/#/docs/main/stable/class/ShardClientUtil?scrollTo=fetchClientValues) or [`broadcastEval`](https://discord.js.org/#/docs/main/stable/class/ShardClientUtil?scrollTo=broadcastEval) to get information from across shards.
```

Reviewer 2: Actually, now that I think of it, you made dedicated sections for both methods. You should link to them, but I'm not sure how to do so with Gitbook. ``[`fetchClientValues`](#FetchClientValues)`` perhaps?

These two functions are your go-to for getting any information from other shards, so get familiar with them!

## FetchClientValues

[`fetchClientValues`](https://discord.js.org/#/docs/main/master/class/ShardClientUtil?scrollTo=fetchClientValues) gets Client properties from all shards. This is what you should use when you would like to get any of the nested properties of the Client, such as `guilds.size` or `uptime`. It's useful for getting things like Collection sizes, basic client properties, and unprocessed information about the client.

Example:
```javascript
/*
    Example result of fetchClientValues() on a bot with 4,300 guilds split across 4 shards.
    Assume this is being executed on shard 0, the first shard.
*/

// If we just get our client.guilds.cache.size, it will return
// only the number of guilds on the shard this is being run on.
console.log('client.guilds.cache.size');
// 1050

// If we would like to get our client.guilds.cache.size from all
// of our shards, we must make use of fetchClientValues().
const res = await client.shard.fetchClientValues('guilds.cache.size');
console.log(res);

> // 	Array: [

Reviewer 2: ```suggestion
// [
```

Reviewer 2: Same thing for all other instances.

//		1050,	// shard 0
//		1100,	// shard 1
//		1075,	// shard 2
//		1075	// shard 3
//	]

````

Let's say you want to do something like get your total server count - In a non-sharded environment, this would be as simple as getting the `client.guilds.cache.size`. However in a sharded environment, `client.guilds.cache.size` will return not the total servers your bot is in. Instead it returns only the total number of servers *on this shard*, like in the first part of the example above.

Here's an example of a function that uses `fetchClientValues()` to first get, then add the total number of guilds from *all shards* (i.e. your bot's total guild count):

```javascript
/*
  	Example by ZiNc#2032
  	The following code fetches total combined shards' server counts.
  
  	discord.js version 12.x
  	client = new discordjs.Client()

>   	returns Number

Reviewer 2: `Number` would refer to the boxed number object instance, it's better to use `number` instead, as it refers to the primitive value instead.

```suggestion
  	returns number
```
